provisioner/aws-manager.js

```diff
@@ -0,0 +1,696 @@
+        Values: ['running', 'pending']
+      }
+    ]}),
+    that.ec2.describeSpotInstanceRequests({
+      Filters: [{
+        Name: 'launch.key-name',
+        Values: [that.keyPrefix + '*']
+      }, {
+        Name: 'state',
+        Values: ['open']
+      }]
+    }),
+  ]);
+
+  p = p.then(function(res) {
+    that.__apiState = that._classify(res[0], res[1]);
+  });
+
+  p = p.then(function() {
+    // We want to make sure that our internal state is always up to date when
+    // we fetch the updated state
+    that.reconcileInternalState();
+  });
+
+  return p;
+};
+
+
+/**
+ * Classify the state received from AWS into something in the shape:
+ * {
+ *   region: {
+ *     workerTypeName: {
+ *       running: [<Instance>],
+ *       pending: [<Instance>],
+ *       spotReq: [<SpotRequest>],
+ *     }
+ *   }
+ * }
+ * The Instance and SpotRequest objects are those returned by AWS.
+ * We flatten the Reservations because we don't really care about that
+ * feature right now.
+ */
+AwsManager.prototype._classify = function(instanceState, spotReqs) {
+  var that = this;
+  var state = {};
+
+  that.ec2.regions.forEach(function(region) {
+    var rState = state[region] = {};
+
+    function x(type) {
+      if (!rState[type]) {
+        rState[type] = {
+          running: [],
+          pending: [],
+          spotReq: [],
+        };
+      }
+    }
+
+    instanceState[region].Reservations.forEach(function(reservation) {
+      reservation.Instances.forEach(function(instance) {
+        var workerType = instance.KeyName.substr(that.keyPrefix.length);
+        x(workerType);
+        rState[workerType][instance.State.Name].push(instance); 
+      });
+    });
+
+    spotReqs[region].SpotInstanceRequests.forEach(function(request) {
+      var workerType = request.LaunchSpecification.KeyName.substr(that.keyPrefix.length);
+    }); 
+
+  });
+
+  return state;
+};
+
+
+/**
+ * Get Api State Only
+ */
+AwsManager.prototype.getApi = function(region, type) {
+  if (region && type) {
+    if (!this.__apiState[region]) {
+      return {};
+    }
+    return this.__apiState[region][type];
+  } else if (region && !type) {
+    return this.__apiState[region];
+  } else if (!region && !type) {
+    return this.__apiState;
+  }
+};
+
+/**
+ * Get Internal State Only.  Remember that Internal state
+ * contains an extra Container object which contains
+ * the the raw `request`, the `workerType` name, the `bid`
+ * and a datetime of when it was `submitted`
+ */
+AwsManager.prototype.getInternal = function(region, type) {
+  if (region && type) {
+    if (!this.__internalState[region]) {
+      return {};
+    }
+    return this.__internalState[region][type];
+  } else if (region && !type) {
+    return this.__internalState[region];
+  } else if (!region && !type) {
+    return this.__internalState;
+  }
+};
+
+
+/**
+ * List the types known in a given region
+ */
+AwsManager.prototype.typesForRegion = function(region) {
```

Comment:
rename to `workerTypesForRegion`... If I understand it right... I thought this was instance types.. needless to say I was confused :)